src/Simulation/Native/src/simulator/wavefunction.hpp

```diff
@@ -10,138 +10,305 @@
+    {
+        gates_.clear();
+        set_qids(other.get_qids());
+        append_gates(other.get_gates());
+    }
 
-                inline std::size_t set_register(const std::vector<unsigned>& qs, std::size_t qmask, std::size_t basis_state, std::size_t original = 0ull)
-                {
-                    std::size_t result = original & ~qmask;
-                    for (unsigned i = 0; i < qs.size(); ++i)
-                        result |= ((basis_state >> i) & 1) << qs[i];
-                    return result;
+    size_t size()
+    {
+        return gates_.size();
+    }
+
+    // Greedy method that finds next appropriate cluster
+    // TODO: it would be nice to provide a comment, outlining the algorithm...
+    std::pair<Cluster, std::vector<logical_qubit_id>> next_cluster(std::list<Cluster>& nextClusters, unsigned maxWidth)
+    {
+        std::vector<logical_qubit_id> myUnion;                            // My qubits touched + Next qubits touched
+        std::vector<logical_qubit_id> myDiff;                             // New qubits touched by Next
+        std::vector<logical_qubit_id> myInter;                            // Old qubits touched by Next
+        std::vector<logical_qubit_id> allInter;                           // My qubits + All touched qubits
+        std::set<logical_qubit_id> myTouched(qids_.begin(), qids_.end()); // My qubits touched
+        std::set<logical_qubit_id> allTouched = myTouched;                // All the qubits touched so far
+
+        for (auto it = nextClusters.rbegin(); it != nextClusters.rend(); ++it)
+        {                                            // Look at the clusters that follow us
+            auto nextQs = it->get_qids();            // Pull off one future cluster
+            std::sort(nextQs.begin(), nextQs.end()); // Has to be sorted for set operations
+            myUnion.clear();
+            std::set_union(
+                nextQs.begin(), nextQs.end(), // See what qubits we and the future cluster touch
+                myTouched.begin(), myTouched.end(), std::back_inserter(myUnion));
+            if (myUnion.size() <= maxWidth)
+            { // It's a candiate if it's not beyond our allowed width
+                myDiff.clear();
+                std::set_difference(
+                    nextQs.begin(), nextQs.end(), // Figure out if any of the future qubits aren't already seen by us
+                    myTouched.begin(), myTouched.end(), std::back_inserter(myDiff));
+                allInter.clear();
+                std::set_intersection(
+                    myDiff.begin(), myDiff.end(), // These are any new qubits that might have already been touched
+                    allTouched.begin(), allTouched.end(), std::back_inserter(allInter));
+                if (allInter.size() == 0)
+                { // If the new qubits are untouched... then this is allowed
+                    auto ret = std::make_pair(std::move(*it), myUnion);
```

Comment:
This comment got lost in the refactor. I'm not sure if the word "Remove" is still appropriate on this specific line, since the erase happens on the line below, but I think preserving the idea will be helpful.
```suggestion
                    auto ret = std::make_pair(std::move(*it), myUnion); // Remove the future cluster and add it our cluster.
```